074-connector-offsets-support.md

```diff
@@ -0,0 +1,180 @@
+* Resetting the offsets for a connector while iterating rapidly in a development environment (so that the connector does not have to be renamed each time)
+* Viewing the in-memory offsets for source connectors on the cluster in order to recover from accidental deletion of the Connect source offsets topic (this is currently possible with the config topic, and can be a life-saver)
+* Monitoring the progress of running connectors on a per-source-partition basis 
+* Skipping records that cause issues with a connector and cannot be addressed using existing error-handling features
+
+## Proposal
+
+Strimzi should include mechanisms to let users manage offsets when utilising the KafkaConnector resources.
+This proposal suggests different approaches for each action, since they are all slightly different.
+In summary:
+ - listing offsets via the KafkaConnector status and enabled/disabled through a property
+ - altering offsets via a new resource called KafkaConnectorOffsets
+ - deleting offsets via an annotation on the KafkaConnector resource
+
+### Request/Response from the API endpoints
+
+The request/response used when interacting with the Kafka Connect API is different for source and sink connectors.
+This is because sink connectors use regular Kafka offsets, whereas source connectors determine their own structure for offsets.
+This difference is reflected in the approaches outlined in this proposal.
+
+Source connector format is:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        // Connector-defined source partition
+      },
+      "offset": {
+        // Connector-defined source offset
+      }
+    }
+  ]
+}
+```
+
+Sink connector format is:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "kafka_topic": // Kafka topic
+        "kafka_partition": // Kafka partition
+      },
+      "offset": {
+        "kafka_offset": // Kafka offset
+      }
+    }
+  ]
+}
+```
+
+### Listing offsets
+The KafkaConnector resource will be updated with a new property to enable the current offsets to be listed in the resource status.
+
+The new property will be a boolean property called `showOffsets` and default to `false`. It should default to `false` because enabling it requires the operator to do an additional API call every time it reconciles the connector.
+If there are a lot of connectors running in the cluster this is a lot of additional API calls.
+
+When the property is set to `true` the status will include the offsets.
+
+The details in the status will be similar to the format returned by the Connect API:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaConnector
+#...
+status:
+  #...
+  connectorOffsets:
+    lastUpdateTime: "2024-05-21T09:08:58.467712198Z"
+    source:
+      - partition: "{\"connector-determined-partition\": \"1\"}" # value is a String containing JSON as received from the API
+        offset: "{\"connector-determined-offset\": \"1\"}" # value is a String containing JSON as received from the API
```

Comment:
The structure of the partition and offset is determined by the connector you are running, so yeah it's a completely random structure from the POV of Strimzi. There's no way to predetermine what it would be. In terms of size I guess it's limited to at least the size of a Kafka record, since it is stored in Kafka, but otherwise there is no limitation. I think the CM idea is a good alternative, I'll have a look at incorporating that into the proposal.

Reply:
Just keep in mind that CMs still have size limits. their main advantage would be that:
* If needed you can probably compress the JSON
* If you are too big, you would get a nicer error to share in the actual resource status and tell the user to disable the offset collection

But the size limit will still be there, so if there is a concern that it will be a common issue, it might be still not sufficient (I have no idea how much of a niche situation that would be to be honest).